### PR TITLE
Fix lattice pruning for vectorized segments

### PR DESCRIPTION
**Pull request by Hespe**
## Description

The current implementation of some lattice pruning methods implemented in `Segment` is broken for vectorized segments.

## Motivation and Context

- [x] I have raised an issue to propose this change (#250)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [x] I have reformatted the code and checked that formatting passes (**required**).
- [x] I have have fixed all issues found by `flake8` (**required**).
- [x] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [x] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

**Comment by Hespe** [comment]
The test failure stems from `BPM` having no `length` parameter, but inheriting a `length` property with default shape `(1,)` from `Element`

File: tests/test_speed_optimizations.py [hunk 1]
```diff
@@ -184,6 +190,25 @@ def test_inactive_magnet_drift_replacement_dtype(dtype: torch.dtype):
...
+            cheetah.Quadrupole(
+                length=torch.tensor([0.2]), k1=torch.tensor([0.0]), name="quad"
+            ),
+            cheetah.Drift(length=torch.tensor([0.4])),
+        ]
+    )
+
+    optimized_segment = segment.inactive_elements_as_drifts(except_for=["quad"])
+
+    assert isinstance(optimized_segment.elements[1], cheetah.Quadrupole)
```
Comment (jank324) [inline tests/test_speed_optimizations.py:209]:
```suggestion
def test_inactive_magnet_drift_except_for():
    """
    Test that an inactive magnet is not replaced by a drift when it is included in the list of exceptions.
    """
    segment = cheetah.Segment(
        elements=[
            cheetah.Drift(length=torch.tensor(0.6)),
            cheetah.Quadrupole(
                length=torch.tensor(0.2), k1=torch.tensor(0.0), name="quad"
            ),
            cheetah.Drift(length=torch.tensor(0.4)),
        ]
    )

    optimized_segment = segment.inactive_elements_as_drifts(except_for=["quad"])

    assert isinstance(optimized_segment.elements[1], cheetah.Quadrupole)
```

File: tests/test_speed_optimizations.py [hunk 2]
```diff
@@ -214,3 +239,39 @@ def test_skippable_elements_reset():
...
+    )
+
+    pruned = segment.without_inactive_zero_length_elements()
+    pruned_except = segment.without_inactive_zero_length_elements(except_for=["dipole"])
+
+    assert len(segment.elements) == 6
+    assert len(pruned.elements) == 3
+    assert len(pruned_except.elements) == 4
+    assert np.allclose(segment.length, pruned.length)
+    assert np.allclose(segment.length, pruned_except.length)
```
Comment (jank324) [inline tests/test_speed_optimizations.py:277]:
```suggestion
def test_without_zerolength_elements():
    """Test that zerolength elements are properly recognized and removed."""
    segment = cheetah.Segment(
        elements=[
            cheetah.Drift(length=torch.tensor([1.0, 2.0])),
            cheetah.Dipole(
                length=torch.tensor(0.0),
                angle=torch.tensor(0.0)
            ),
            cheetah.Dipole(
                length=torch.tensor(0.0),
                angle=torch.tensor(0.0),
                name="dipole"
            ),
            cheetah.Dipole(
                length=torch.tensor([0.0, 0.1]),
                angle=torch.tensor(0.0)
            ),
            cheetah.Drift(length=torch.tensor(0.0)),
            cheetah.Dipole(
                length=torch.tensor(0.0),
                angle=torch.tensor([0.5, 0.0])
            ),
        ]
    )

    pruned = segment.without_inactive_zero_length_elements()
    pruned_except = segment.without_inactive_zero_length_elements(except_for=["dipole"])

    assert len(segment.elements) == 6
    assert len(pruned.elements) == 3
    assert len(pruned_except.elements) == 4
    assert np.allclose(segment.length, pruned.length)
    assert np.allclose(segment.length, pruned_except.length)
```

File: tests/test_speed_optimizations.py [hunk 3]
```diff
@@ -184,6 +190,25 @@ def test_inactive_magnet_drift_replacement_dtype(dtype: torch.dtype):
...
+        elements=[
+            cheetah.Drift(length=torch.tensor(0.6)),
+            cheetah.Quadrupole(
+                length=torch.tensor(0.2), k1=torch.tensor(0.0), name="quad"
+            ),
+            cheetah.Drift(length=torch.tensor(0.4)),
+        ]
+    )
+
+    optimized_segment = segment.inactive_elements_as_drifts(except_for=["quad"])
```
Comment (jank324) [inline tests/test_speed_optimizations.py:207]:
```suggestion
                length=torch.tensor(0.2), k1=torch.tensor(0.0), name="my_quad"
            ),
            cheetah.Drift(length=torch.tensor(0.4)),
        ]
    )

    optimized_segment = segment.inactive_elements_as_drifts(except_for=["my_quad"])
```

File: tests/test_speed_optimizations.py [hunk 4]
```diff
@@ -214,3 +239,39 @@ def test_skippable_elements_reset():
...
+            cheetah.Drift(length=torch.tensor(0.0)),
+            cheetah.Dipole(
+                length=torch.tensor(0.0),
+                angle=torch.tensor([0.5, 0.0])
+            ),
+        ]
+    )
+
+    pruned = segment.without_inactive_zero_length_elements()
+    pruned_except = segment.without_inactive_zero_length_elements(except_for=["dipole"])
```
Comment (jank324) [inline tests/test_speed_optimizations.py:271]:
```suggestion
                name="my_dipole"
            ),
            cheetah.Dipole(
                length=torch.tensor([0.0, 0.1]),
                angle=torch.tensor(0.0)
            ),
            cheetah.Drift(length=torch.tensor(0.0)),
            cheetah.Dipole(
                length=torch.tensor(0.0),
                angle=torch.tensor([0.5, 0.0])
            ),
        ]
    )

    pruned = segment.without_inactive_zero_length_elements()
    pruned_except = segment.without_inactive_zero_length_elements(except_for=["my_dipole"])
```

**Comment by jank324** [comment]
@Hespe, I tried adapting the tests to the new broadcasting semantics. Can you check that in doing so I didn't invalidate their purpose?

**Comment by Hespe** [comment]
I did some further cleanup on the tests. Looks good to me now.

**Review by jank324 — APPROVED** [review]

**Pull request merged**